flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvSerDeSchemaFactory.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.csv;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.format.DecodingFormat;
+import org.apache.flink.table.connector.format.EncodingFormat;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.factories.DeserializationFormatFactory;
+import org.apache.flink.table.factories.DynamicTableFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+import org.apache.flink.table.factories.SerializationFormatFactory;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.RowType;
+
+import org.apache.commons.lang3.StringEscapeUtils;
+
+import java.util.Collections;
+import java.util.Set;
+
+import static org.apache.flink.formats.csv.CsvFormatOptions.ALLOW_COMMENTS;
+import static org.apache.flink.formats.csv.CsvFormatOptions.ARRAY_ELEMENT_DELIMITER;
+import static org.apache.flink.formats.csv.CsvFormatOptions.DISABLE_QUOTE_CHARACTER;
+import static org.apache.flink.formats.csv.CsvFormatOptions.ESCAPE_CHARACTER;
+import static org.apache.flink.formats.csv.CsvFormatOptions.FIELD_DELIMITER;
+import static org.apache.flink.formats.csv.CsvFormatOptions.IGNORE_PARSE_ERRORS;
+import static org.apache.flink.formats.csv.CsvFormatOptions.NULL_LITERAL;
+import static org.apache.flink.formats.csv.CsvFormatOptions.QUOTE_CHARACTER;
+
+/**
+ * Format factory for providing configured instances of CSV to RowData {@link SerializationSchema}
+ * and {@link DeserializationSchema}.
+ */
+@Internal
+public final class CsvSerDeSchemaFactory
+        implements DeserializationFormatFactory, SerializationFormatFactory {
+
+    @Override
+    public DecodingFormat<DeserializationSchema<RowData>> createDecodingFormat(
+            DynamicTableFactory.Context context, ReadableConfig formatOptions) {
+        FactoryUtil.validateFactoryOptions(this, formatOptions);
+        CsvCommons.validateFormatOptions(formatOptions);
+
+        return new DecodingFormat<DeserializationSchema<RowData>>() {
+            @Override
+            public DeserializationSchema<RowData> createRuntimeDecoder(
```

Comment:
I wonder if this one can be a projectable decoding format as well, given it roughly uses the same code for deserializing.

Reply:
I guess it could. 
@AHeise, I believe we did not previously discuss this aspect. Do you think it is necessary to expand the scope of projection pushdown support also to the Ser/De schema-based implementation for this PR?

Reply:
This would be orthogonal and can be done later. I'm also not sure if it's worth the hassle now: performance gain on a per-record projection is comparably small and we implicitly have it on all formats with reader schema.

However, I'd like you to verify if the current version actually works with projections, so can you please add an appropriate test?

Reply:
@AHeise But then why `CsvFormatFactory` returns a `ProjectableDecodingFormat`? It doesn't make sense, becuse either none of them are projectable or both are.

Reply:
Is this fully supported in Table-land? If so, then I agree we should change it. It's a good performance-improvement.

Reply:
> Is this fully supported in Table-land? If so, then I agree we should change it. It's a good performance-improvement.

Yes, please check the documentation of `DecodingFormat` which goes through all the implementation details both for formats and connectors

Reply:
@slinkydeveloper 
I am not sure about this actually. The focus of this PR is 
**Add CSV format support for filesystem based on StreamFormat and BulkWriter interfaces.**
I would rather not touch the existing Ser/De schema-based functionality in this scope. Nothing will break if we won't support pushdown in the non-filesystem connectors for now and to me, it sounds like an improvement that should be done in a separate PR.

Reply:
Notice that `CsvSerDeSchemaFactory` is just a rename of the previously existing `CsvFormatFactory`, we did not touch its functionality in the PR.

Reply:
Ok but then open a followup issue for that, because this inconsistency between the `CsvFormatFactory` and `CsvSerDeSchemaFactory` is confusing and doesn't really make sense to exist, as either both support projections, or none of them does.

Reply:
@slinkydeveloper as discussed, https://issues.apache.org/jira/browse/FLINK-25931.